diff --git a/CPP/7zip/UI/Common/Update.cpp b/CPP/7zip/UI/Common/Update.cpp
index 92c03352..c8834571 100644
--- a/CPP/7zip/UI/Common/Update.cpp
+++ b/CPP/7zip/UI/Common/Update.cpp
@@ -1027,6 +1027,7 @@ static HRESULT EnumerateInArchiveItems(
   return S_OK;
 }
 
+#if 0
 #if defined(_WIN32) && !defined(UNDER_CE)
 
 #if defined(__MINGW32__) || defined(__MINGW64__)
@@ -1034,6 +1035,7 @@ static HRESULT EnumerateInArchiveItems(
 #else
 #include <MAPI.h>
 #endif
+#endif
 
 extern "C" {
 
@@ -1354,9 +1356,11 @@ HRESULT UpdateArchive(
 
       dirItems.SymLinks = options.SymLinks.Val;
 
+	  #if 0
       #if defined(_WIN32) && !defined(UNDER_CE)
       dirItems.ReadSecure = options.NtSecurity.Val;
       #endif
+	  #endif
 
       dirItems.ScanAltStreams = options.AltStreams.Val;
       dirItems.ExcludeDirItems = censor.ExcludeDirItems;
@@ -1624,7 +1628,7 @@ HRESULT UpdateArchive(
     }
   }
 
-
+  #if 0
   #if defined(_WIN32) && !defined(UNDER_CE)
   
   if (options.EMailMode)
@@ -1774,6 +1778,7 @@ HRESULT UpdateArchive(
    }
   }
   
+  #endif
   #endif
 
   if (options.DeleteAfterCompressing)
